cpp/src/arrow/util/decimal_test.cc

```diff
@@ -641,13 +698,79 @@ TEST(TestDecimalFromRealDouble, LargeValues) {
+  }
+}
+
+// Same tests for Decimal256
+class TestDecimal256FromRealDouble
+    : public ::testing::TestWithParam<FromDoubleTestParam> {};
+
+TEST_P(TestDecimal256FromRealDouble, SuccessConversion) {
+  const auto param = GetParam();
+  CheckDecimalFromReal<Decimal256>(param.real, param.precision, param.scale,
+                                   param.expected);
+}
+
+// clang-format off
+INSTANTIATE_TEST_SUITE_P(
+    TestDecimal256FromRealDouble, TestDecimal256FromRealDouble,
+    ::testing::Values(
+        // 2**63 + 2**11 (exactly representable in a double's 53 bits of precision)
+        FromDoubleTestParam{9.223372036854778e+18, 19, 0, "9223372036854777856"},
+        FromDoubleTestParam{-9.223372036854778e+18, 19, 0, "-9223372036854777856"},
+        FromDoubleTestParam{9.223372036854778e+10, 19, 8, "92233720368.54777856"},
+        FromDoubleTestParam{-9.223372036854778e+10, 19, 8, "-92233720368.54777856"},
+        // 2**64 - 2**11 (exactly representable in a double)
+        FromDoubleTestParam{1.844674407370955e+19, 20, 0, "18446744073709549568"},
+        FromDoubleTestParam{-1.844674407370955e+19, 20, 0, "-18446744073709549568"},
+        // 2**64 + 2**11 (exactly representable in a double)
+        FromDoubleTestParam{1.8446744073709556e+19, 20, 0, "18446744073709555712"},
+        FromDoubleTestParam{-1.8446744073709556e+19, 20, 0, "-18446744073709555712"},
+        FromDoubleTestParam{1.8446744073709556e+15, 20, 4, "1844674407370955.5712"},
+        FromDoubleTestParam{-1.8446744073709556e+15, 20, 4, "-1844674407370955.5712"},
+        // Almost 10**38 (minus 2**73)
+        FromDoubleTestParam{9.999999999999998e+37, 38, 0,
+                            "99999999999999978859343891977453174784"},
+        FromDoubleTestParam{-9.999999999999998e+37, 38, 0,
+                            "-99999999999999978859343891977453174784"},
+        FromDoubleTestParam{9.999999999999998e+27, 38, 10,
+                            "9999999999999997885934389197.7453174784"},
+        FromDoubleTestParam{-9.999999999999998e+27, 38, 10,
+                            "-9999999999999997885934389197.7453174784"},
+        // Almost 10**76
+        FromDoubleTestParam{9.999999999999999e+75, 76, 0,
+                            "999999999999999886366330070006442034959750906670402"
+                            "8242075715752105414230016"},
+        FromDoubleTestParam{-9.999999999999999e+75, 76, 0,
+                            "-999999999999999886366330070006442034959750906670402"
+                            "8242075715752105414230016"},
+        FromDoubleTestParam{9.999999999999999e+65, 76, 10,
+                            "999999999999999886366330070006442034959750906670402"
+                            "824207571575210.5414230016"},
+        FromDoubleTestParam{-9.999999999999999e+65, 76, 10,
+                            "-999999999999999886366330070006442034959750906670402"
+                            "824207571575210.5414230016"}
```

Comment:
This is the same test vector as above + 4 additional values. Perhaps factor out the common values?